### PR TITLE
feat(s3): introduce S3 connector and action

### PR DESCRIPTION
**Pull request by keynslug**
Fixes [EMQX-11830](https://emqx.atlassian.net/browse/EMQX-11830).

Release version: e5.6

## Summary

This is a trivial connector based on `emqx_s3` and simple action that maps each incoming event into an S3 object upload. Due to the current `emqx_s3` limitation this bridge is compatible with backends providing S3 API with path-style bucket access.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] ~~Added property-based tests for code which performs user input validation~~
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [x] Schema changes are backward compatible

### CODE REVIEW

File: apps/emqx_bridge_s3/src/emqx_bridge_s3.erl [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    lists:append([
+        emqx_connector_schema:common_fields(),
+        fields(s3_connector_config),
+        [
+            {resource_opts,
+                hoconsc:mk(
+                    ?R_REF(s3_connector_resource_opts),
+                    emqx_resource_schema:resource_opts_meta()
+                )}
+        ]
```
Comment (thalesmg) [inline apps/emqx_bridge_s3/src/emqx_bridge_s3.erl:65]: There's also a convenience function `emqx_connector_schema:resource_opts_ref` that could be used.

File: apps/emqx_bridge_s3/src/emqx_bridge_s3_connector.erl [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+}.
+
+-type channel_state() :: #{
+    bucket := emqx_template:str(),
+    key := emqx_template:str(),
+    upload_options => emqx_s3_client:upload_options()
+}.
+
+-type state() :: #{
+    pool_name => resource_id(),
```
Comment (thalesmg) [inline apps/emqx_bridge_s3/src/emqx_bridge_s3_connector.erl:49]: Tiny nit:

```suggestion
    pool_name := resource_id(),
```

**Review by thalesmg — APPROVED** [review]

File: apps/emqx_bridge_s3/src/emqx_bridge_s3_connector.erl [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    State = #{
+        pool_name => PoolName,
+        client_config => emqx_s3_profile_conf:client_config(S3Config, PoolName),
+        channels => #{}
+    },
+    HttpConfig = emqx_s3_profile_conf:http_config(Config),
+    case ehttpc_sup:start_pool(PoolName, HttpConfig) of
+        {ok, Pid} ->
+            ?SLOG(info, #{msg => "s3_connector_start_http_pool_success", pool_name => PoolName}),
+            {ok, State#{pool_pid => Pid}};
```
[the remark this answers is not in the document]
Comment (keynslug) [reply apps/emqx_bridge_s3/src/emqx_bridge_s3_connector.erl:77]: Yes. Decided to keep it for introspection purposes, sort of.

**Review by thalesmg — DISMISSED** [review]

**Review by savonarola — DISMISSED** [review]

File: apps/emqx_bridge_s3/src/emqx_bridge_s3_action_info.erl [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+-export([
+    action_type_name/0,
+    connector_type_name/0,
+    schema_module/0
+]).
+
+action_type_name() -> s3_upload.
+
+connector_type_name() -> s3.
```
Comment (zmstone) [inline apps/emqx_bridge_s3/src/emqx_bridge_s3_action_info.erl:17]: I remember there is a implicit contract with cloud UI team that action type must match connector type.

File: rel/i18n/emqx_s3_schema.hocon [hunk 5]
```diff
@@ -9,6 +9,9 @@ secret_access_key.desc:
 bucket.desc:
 """The name of the S3 bucket."""
 
+key.desc:
+"""Key of the S3 object being manipulated."""
```
Comment (zmstone) [inline rel/i18n/emqx_s3_schema.hocon:13]:
```suggestion
"""Key of the S3 object."""

key.label:
"""Object Key"""
```

File: rel/i18n/emqx_s3_schema.hocon [hunk 6]
```diff
@@ -9,6 +9,9 @@ secret_access_key.desc:
 bucket.desc:
 """The name of the S3 bucket."""
```
Comment (zmstone) [inline rel/i18n/emqx_s3_schema.hocon:10]:
```suggestion
"""The name of the S3 bucket."""

bucket.label:
"""Bucket"""
```

File: apps/emqx_s3/src/emqx_s3_client.erl [hunk 7]
```diff
@@ -273,10 +278,10 @@ request_fun(HttpPool, PoolType, MaxRetries) ->
         end)
     end.
 
-ehttpc_request(HttpPool, Method, Request, Timeout, MaxRetries) ->
-    try timer:tc(fun() -> ehttpc:request(HttpPool, Method, Request, Timeout, MaxRetries) end) of
+ehttpc_request(Worker, Method, Request, Timeout, MaxRetries) ->
```
Comment (zmstone) [inline apps/emqx_s3/src/emqx_s3_client.erl:281]: Q: given that we inject a http request func from config, make sense to delete all the http client dependencies from our erlcloud fork ?

File: apps/emqx_bridge_s3/src/emqx_bridge_s3_action_info.erl (hunk 4 above)
Comment (keynslug) [reply apps/emqx_bridge_s3/src/emqx_bridge_s3_action_info.erl:17]: Interesting. I'll change them to be identical, sort of curios now though: what if we want to have an application with 1 connector and, let's say, 3 different actions (e.g. simple upload, assembled upload, metadata update)?

File: apps/emqx_s3/src/emqx_s3_client.erl (hunk 7 above)
Comment (keynslug) [reply apps/emqx_s3/src/emqx_s3_client.erl:281]: I think we still use the default client (hackney) in `emqx_bridge_dynamo` and in tests. Besides, `ehttpc_pool`-based approach won't currently work smoothly if we'd want support for vhost-only bucket access, supposedly the easiest way would be to switch back to hackney (though not the only way).

Comment (zmstone) [reply apps/emqx_s3/src/emqx_s3_client.erl:281]: ok, at least lhttpc is not needed https://github.com/emqx/erlcloud/blob/2e95901cfd434e05ac6871d5a3134dd757b2cefd/rebar.config#L21 ?

**Review by zmstone — APPROVED** [review]

**Pull request merged**